docs/lazy-adr/adr-009-state-fraud-proofs.md

```diff
@@ -0,0 +1,233 @@
+}
+
+message RequestGetAppHash {}
+
+message ResponseGetAppHash {
+  bytes app_hash = 1;
+}
+```
+
+This `GetAppHash` ABCI method returns an equivalent of `CommitID` hash in the ABCI method `Commit` and thus provides a way to extract ISRs from an app without doing any disk write operations.
+
+Full nodes use these ISRs to detect fraudulent state transitions. A full node must also execute all state transitions (`BeginBlock`, `DeliverTx`, and `EndBlock` calls) and compute its own Intermediate State Roots (ISRs). After each state transition, a full node compares the corresponding ISR with the ISR given by the Sequencer. If it finds a mismatch between its own computed ISR and and one given by the Sequencer, a fraudulent transition is detected and it moves on to generate a Fraud Proof.
+
+### Generating Fraud Proofs
+
+We introduce the following ABCI method to enable Fraud Proof Generation in the Cosmos SDK:
+
+```protobuf
+service ABCIApplication {
+    rpc GenerateFraudProof(RequestGenerateFraudProof) returns (ResponseGenerateFraudProof);
+}
+```
+
+With this new ABCI method, a Rollmint Full Node can send a request to a Cosmos SDK app to generate a Fraud Proof. In this request, it includes a list of all the state transitions from the start of the block upto the fraudulent state transition. The last non-nil state transition in this list corresponds to the fraudulent state transition.
+
+```protobuf
+message RequestGenerateFraudProof {
+  // All state transitions upto and including the fraudulent state transition
+  RequestBeginBlock begin_block_request = 1 [(gogoproto.nullable) = false];
+  repeated RequestDeliverTx deliver_tx_requests = 2;
+  RequestEndBlock end_block_request = 3;
+}
+```
+
+The `GenerateFraudProof` method in the Cosmos SDK app receives this list of state transitions and takes the following steps to generate a fraud proof:
+
+- Revert local state to the last committed state
+- Execute all the non-fraudulent state transitions
+- Remove all trace logs from executing the above state transitions
+- Execute the fraudulent state transition with tracing-enabled and store logs of what state, specifically key/value pairs, is accessed, modified, added, or removed during this fraudulent state transition.
+- Revert local state back to the last committed state
+- Execute all the non-fraudulent state transitions again
+- Use the stored logs to filter the local state down to the set of minimal set of key/value pairs needed to execute the fraudulent state transition and generate corresponding merkle inclusion proofs for the state.
+- Put this minimal set of key/value pairs and corresponding merkle inclusion proofs in a State Fraud Proof which looks like this:
+
+```protobuf
+
+// Represents a single-round fraudProof
+message FraudProof {
+  // The block height during which the fraudulent state transition occured
+  int64 block_height = 1;
+  // Intermediate State Root right before the fraudulent state transition
+  bytes app_hash = 2;
```

Comment:
As per my prior comments - it needs to be verified that the app hash is part of the rollup's data commitment, otherwise fraud proof generators could make up fake ISRs and transactions.

Reply:
We keep this in mind. But first we need to complete https://github.com/celestiaorg/rollmint/issues/594 and 
https://github.com/celestiaorg/rollmint/issues/613.